File: experiments/source/python/huber_sum.py
import loom


@loom.entry
def huber_sum(delta: float, x: loom.Tensor1, y: loom.Tensor1) -> float:
    def contribution(xi: float, yi: float) -> float:
        d = xi - yi
        return (
            delta * (d - (0.5 * delta))
            if d > delta
            else delta * ((-d) - (0.5 * delta))
            if d < (-delta)
            else 0.5 * d * d
        )

    return loom.Tensor1.reduce_sum(loom.Tensor1.map2(contribution, x, y))

File: experiments/source/cpp/huber_sum.cpp
#include <loom/loom.hpp>

LOOM_ENTRY float huber_sum(float delta, loom::Tensor1 x, loom::Tensor1 y) {
  auto contribution = [=](float xi, float yi) -> float {
    auto d = xi - yi;
    return d > delta ? delta * (d - (0.5f * delta))
         : d < (-delta) ? delta * ((-d) - (0.5f * delta))
         : 0.5f * d * d;
  };
  return loom::Tensor1::reduce_sum(loom::Tensor1::map2(contribution, x, y));
}
